File: src/main/kotlin/io/xhub/confIOMessenger/domain/Message.java
package io.xhub.confIOMessenger.domain;

import com.fasterxml.jackson.annotation.JsonProperty;

import java.io.Serializable;

public class Message<T extends AbstractMessageBody> implements Serializable {
    @JsonProperty("messaging_type")
    private String messagingType;
    private Identifiable sender;
    private Identifiable recipient;
    private T message;
    private Long timestamp;

    public String getMessagingType() {
        return messagingType;
    }

    public void setMessagingType(String messagingType) {
        this.messagingType = messagingType;
    }

    public Identifiable getSender() {
        return sender;
    }

    public void setSender(Identifiable sender) {
        this.sender = sender;
    }

    public Identifiable getRecipient() {
        return recipient;
    }

    public void setRecipient(Identifiable recipient) {
        this.recipient = recipient;
    }

    public T getMessage() {
        return message;
    }

    public void setMessage(T message) {
        this.message = message;
    }

    public Long getTimestamp() {
        return timestamp;
    }

    public void setTimestamp(Long timestamp) {
        this.timestamp = timestamp;
    }
}

File: src/main/kotlin/io/xhub/confIOMessenger/domain/Message.kt
package io.xhub.confIOMessenger.domain

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.databind.node.BooleanNode
import io.vertx.core.json.JsonObject
import io.xhub.confIOMessenger.annotation.IgnoreNotNullUnknown

@IgnoreNotNullUnknown
sealed class Message(
        @JsonProperty()
        open val text:String?=null,
        @JsonProperty("is_echo")
        val is_echo:Boolean?= null
){
    @IgnoreNotNullUnknown
    data  class ReceivedMessage(
            @JsonProperty("text")
            override var  text:String?=null,
            @JsonProperty("attachments")
            val attachments:List<Attachement>?=null,
            @JsonProperty("quick_reply")
            val quick_reply:QuickReply?=null
    ) :Message(text)
    @IgnoreNotNullUnknown
    data class SentMessage(
            @JsonProperty("text")
            override var text:String?=null,
            @JsonProperty("attachment")
            val attachment:Attachement?=null,
            @JsonProperty("quick_replies")
            val quick_replies:List<QuickReply>?= null
    ):Message(text)
}
